deepmd/utils/tabulate.py

```diff
@@ -161,81 +163,100 @@ def build(self,
                     if self.type_one_side:
                         net = "filter_-1_net_" + str(ii)
+                        # upper and lower should consider all types which are not excluded and sel>0
+                        idx = [(type_i, ii) not in self.exclude_types and self.sel_a[type_i] > 0 for type_i in range(self.ntypes)]
+                        uu = np.max(upper[idx])
+                        ll = np.min(lower[idx])
                     else:
-                        net = "filter_" + str(ii // self.ntypes) + "_net_" + str(ii % self.ntypes)
-                    self._build_lower(net, xx, ii, upper, lower, stride0, stride1, extrapolate)
+                        ielement = ii // self.ntypes
+                        net = "filter_" + str(ielement) + "_net_" + str(ii % self.ntypes)
+                        uu = upper[ielement]
+                        ll = lower[ielement]
+                    xx = np.arange(ll, uu, stride0, dtype = self.data_type)
+                    xx = np.append(xx, np.arange(uu, extrapolate * uu, stride1, dtype = self.data_type))
+                    xx = np.append(xx, np.array([extrapolate * uu], dtype = self.data_type))
+                    nspline = ((uu - ll) / stride0 + (extrapolate * uu - uu) / stride1).astype(int)
+                    self._build_lower(net, xx, ii, uu, ll, stride0, stride1, extrapolate, nspline)
         elif isinstance(self.descrpt, deepmd.descriptor.DescrptSeT):
-            xx = np.arange(extrapolate * lower, lower, stride1, dtype = self.data_type)
-            xx = np.append(xx, np.arange(lower, upper, stride0, dtype = self.data_type))
-            xx = np.append(xx, np.arange(upper, extrapolate * upper, stride1, dtype = self.data_type))
-            xx = np.append(xx, np.array([extrapolate * upper], dtype = self.data_type))
-            self.nspline = int((upper - lower) / stride0 + 2 * ((extrapolate * upper - upper) / stride1))
+            xx_all = []
+            for ii in range(self.ntypes):
+                xx = np.arange(extrapolate * lower[ii], lower[ii], stride1, dtype = self.data_type)
+                xx = np.append(xx, np.arange(lower[ii], upper[ii], stride0, dtype = self.data_type))
+                xx = np.append(xx, np.arange(upper[ii], extrapolate * upper[ii], stride1, dtype = self.data_type))
+                xx = np.append(xx, np.array([extrapolate * upper[ii]], dtype = self.data_type))
+                xx_all.append(xx)
+            nspline = ((upper - lower) / stride0 + 2 * ((extrapolate * upper - upper) / stride1)).astype(int)
             idx = 0
             for ii in range(self.ntypes):
                 for jj in range(ii, self.ntypes):
                     net = "filter_" + str(ii) + "_net_" + str(jj)
-                    self._build_lower(net, xx, idx, upper, lower, stride0, stride1, extrapolate)
+                    self._build_lower(net, xx_all[ii], idx, upper[ii], lower[ii], stride0, stride1, extrapolate, nspline[ii])
```

Comment:
Should it be changed to nspline instead of nspline[ii]?
```suggestion
                    self._build_lower(net, xx_all[ii], idx, upper[ii], lower[ii], stride0, stride1, extrapolate, nspline)
```

Reply:
Here `nspline` is an 1-D array (different from others).